url60801aaae5_yurgenbeerman_JavaOOP_with_Maven_tgz-pJ8-edu_services_PublicServiceDemoJ8/src/edu/clients/Citizen.java

```diff
@@ -0,0 +1,192 @@
+        /* TODO validate passed parameter:
+         *   isEmailAddressValid(emailFromAddress)
+         *   Found two best validators:
+         *     1) http://commons.apache.org/proper/commons-validator/download_validator.cgi
+         *     2) https://java.net/projects/javamail/pages/Home#Download_JavaMail_1.5.1_Release
+         */
+        this.emailAddress = emailAddress;
+    }
+
+    public Address getAddress() {
+        return address;
+    }
+
+    public String getAddressString() {
+        if ( null != address) {
+            return address.toString();
+        } else {
+            throw new IllegalStateException(ClientsDefaults.NO_ADDRESS);
+        }
+    }
+
+    public void setAddress(Address address) {
+        this.address = address;
+    }
+
+    public long getId() {
+        return citizenId;
+    }
+
+    public String getName() {
+        return this.getFullNameString();
+    }
+
+    public String getOfficialId() {
+        return officialId;
+    }
+
+    public void setOfficialId(String officialId) {
+        this.officialId = officialId;
+    }
+
+    public ArrayList<OutcomingDocument> getResponses() {
+        return responses;
+    }
+
+    public String getResponsesString() {
+        String result = "";
+        if ( getResponses() != null ) {
+            for (int i = 0; i < getResponses().size(); i++) {
+                result += getResponses().get(i);
+            }
+        }
+        return result;
+    }
+
+    public void addResponse(OutcomingDocument response) {
+        if (null == getResponses()) {
+            this.responses = new ArrayList<OutcomingDocument>();
+        }
+        this.responses.add(response);
+    }
+
+    public String toString() {
+        return this.getFullName().getName();
+    }
+
+    public boolean equals(Object other) {
+        if (other == this)
+            return true;
+        if (! (other instanceof Citizen) )
+            return false;
+        Citizen otherCitizen = (Citizen) other;
+        return (fullName == null) ?
+                (otherCitizen.fullName == null) :
+        (fullName.equals(otherCitizen.fullName) && officialId.equals(otherCitizen.officialId));
+    }
+
+    public int hashCode() {
+        return (officialId.hashCode() + fullName.hashCode());
+    }
+
+    private static synchronized void setCitizenId(Citizen citizen) {
+        citizen.citizenId = nextCitizenId;
+        Citizen.nextCitizenId++;
+    }
+
+    public long getCitizenId() {
+        return this.citizenId;
+    }
+
+    public int compareTo(Citizen other) {
+        if ( this.getCitizenId() == other.getCitizenId() )
+            return 0;
+        return ( this.getCitizenId() > other.getCitizenId() ) ? 1 : -1;
+    }
+
+    public boolean isRequesterOfficialIdValid(OfficialIDsHolder officialIDsHolder) {
+        DocumentCreator documentCreator = this;
+        if (( documentCreator.getOfficialId() != null) && (documentCreator.getOfficialId() != "")) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.